p2p/src/main/java/bisq/network/p2p/storage/persistence/SplitStoreService.java

```diff
@@ -0,0 +1,201 @@
+
+        // go through the list one by one
+        versions.forEach(version -> {
+            String filename = getFileName() + "_" + version;
+            if (new File(absolutePathOfStorageDir, filename).exists()) {
+                // if it is there already, load
+                history.put(version, readStore(getFileName() + "_" + version));
+            } else {
+                // either copy and split
+                history.put(version, copyAndSplit(version, postFix));
+            }
+        });
+    }
+
+    /**
+     * Bluntly copy and pasted from {@link StoreService} because:
+     * <ul><li>The member function there is private</li>
+     * <li>it does not match our interface</li>
+     * <li>is a temporary solutions, until https://github.com/bisq-network/projects/issues/29</li></ul>
+     *
+     * @param name
+     * @return store
+     */
+    private T readStore(String name) {
+        T store = storage.initAndGetPersistedWithFileName(name, 100);
+        if (store != null) {
+            log.info("{}: size of {}: {} MB", this.getClass().getSimpleName(),
+                    storage.getClass().getSimpleName(),
+                    store.toProtoMessage().toByteArray().length / 1_000_000D);
+        } else {
+            store = createStore();
+        }
+
+        return store;
+    }
+
+    /**
+     * Copy the missing data store from resources and remove its object from the live store.
+     *
+     * @param version to identify the data store eg. "1.3.4"
+     * @param postFix the global postfix eg. "_BTC_MAINNET"
+     * @return the freshly copied and loaded data store
+     */
+    private SplitStore copyAndSplit(String version, String postFix) {
+        // if not, copy and split
+        final File destinationFile = new File(absolutePathOfStorageDir, getFileName() + "_" + version);
+        String resourceFileName = destinationFile.getName() + postFix; // postFix has a preceding "_" already
+        try {
+            log.info("We copy resource to file: resourceFileName={}, destinationFile={}", resourceFileName, destinationFile);
+            FileUtil.resourceToFile(resourceFileName, destinationFile);
+        } catch (ResourceNotFoundException e) {
+            log.info("Could not find resourceFile {}. That is expected if none is provided yet.", resourceFileName);
+        } catch (Throwable e) {
+            log.error("Could not copy resourceFile {} to {}.\n{}", resourceFileName, destinationFile.getAbsolutePath(), e.getMessage());
+            e.printStackTrace();
+        }
+
+        // split
+        // - get all
+        SplitStore historicalStore = readStore(destinationFile.getName());
+        // - subtract all that is in resource files
+        store.getMap().keySet().removeAll(historicalStore.getMap().keySet());
+
+        // - create new file with leftovers
+        storage.initAndGetPersisted(store, 0);
+        storage.queueUpForSave();
+
+        return historicalStore;
+    }
```

Comment:
I see this whole PR a small-as-can-be bugfix for the 3/4 problem. The "followup"-project would be to remove the protobuf-files altogether and make everything work with a proper database (https://github.com/bisq-network/projects/issues/29). When we approach that project, all the file-handling will be removed and cleaned up.

I agree that the implementation is not as clean as could be, but I rather focus my efforts and not do stuff twice.

Reply:
> Why are we copying historical datastore checkpoints/splits out of resources? Because that's what the live-store-only solution we currently use does? For troubleshooting purposes? Seems unnecessary, though I may be missing something. 

the existing file storage mechanisms are all tailored to working in the working directory. Changing that would be equivalent to replacing the file storage mechanisms with something else entirely. Hence, leaving it as is (until https://github.com/bisq-network/projects/issues/29).

> I presume we keep the live store in the working directory, because we can't mutate resource files (is that even true?).

Well, obviously, the "live store" is something that is created and maintained by the Bisq app per installation and even per (configurable) app name and path. Placing that in the resources would only allow for a single instance of the data store AND would be overwritten and therefore invalidated on every app upgrade.